cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -377,24 +382,32 @@ public boolean apply(Id.Program programId) {
    * @param appId the {@link Id.Application} of the application to be removed
    * @throws Exception
    */
-  public void removeApplication(final Id.Application appId) throws Exception {
+  public Set<ProgramId> removeApplication(final Id.Application appId) throws Exception {
```

Comment:
Same here. I guess it's better to throw `CannotBeDeletedException` instead of returning an empty set.